File: CK3toEU4/Source/CK3World/Titles/Title.cpp
#include "Title.h"
#include "../../Mappers/DevWeightsMapper/DevWeightsMapper.h"
#include "../Characters/Character.h"
#include "../Geography/CountyDetail.h"
#include "../Geography/ProvinceHolding.h"
#include "LandedTitles.h"
#include "Log.h"
#include "ParserHelpers.h"

CK3::Title::Title(std::istream& theStream, long long theID): ID(theID)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();

	// Obligatory sanity check.
	if (!holder)
	{
		// This title does NOT in fact exist.
		dfLiege.reset();
	}
}

void CK3::Title::registerKeys()
{
	registerKeyword("key", [this](const std::string& unused, std::istream& theStream) {
		name = commonItems::singleString(theStream).getString();
	});
	registerKeyword("name", [this](const std::string& unused, std::istream& theStream) {
		displayName = commonItems::singleString(theStream).getString();
		if (displayName.find("\x15") != std::string::npos)
		{
			cleanUpDisplayName();
		}
	});
	registerKeyword("adj", [this](const std::string& unused, std::istream& theStream) {
		adjective = commonItems::singleString(theStream).getString();
	});
	registerKeyword("date", [this](const std::string& unused, std::istream& theStream) {
		creationDate = date(commonItems::singleString(theStream).getString());
	});
	registerKeyword("claim", [this](const std::string& unused, std::istream& theStream) {
		for (auto claimantID: commonItems::llongList(theStream).getLlongs())
			claimants.insert(std::make_pair(claimantID, nullptr));
	});
	registerKeyword("history_government", [this](const std::string& unused, std::istream& theStream) {
		historyGovernment = commonItems::singleString(theStream).getString();
	});
	registerKeyword("theocratic_lease", [this](const std::string& unused, std::istream& theStream) {
		theocraticLease = commonItems::singleString(theStream).getString() == "yes";
	});
	registerKeyword("capital_barony", [this](const std::string& unused, std::istream& theStream) {
		cCapitalBarony = commonItems::singleString(theStream).getString() == "yes";
	});
	registerKeyword("duchy_capital_barony", [this](const std::string& unused, std::istream& theStream) {
		dCapitalBarony = commonItems::singleString(theStream).getString() == "yes";
	});
	registerKeyword("capital", [this](const std::string& unused, std::istream& theStream) {
		capital = std::pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("de_facto_liege", [this](const std::string& unused, std::istream& theStream) {
		dfLiege = std::pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("de_jure_liege", [this](const std::string& unused, std::istream& theStream) {
		djLiege = std::pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("de_jure_vassals", [this](const std::string& unused, std::istream& theStream) {
		for (auto vassalID: commonItems::llongList(theStream).getLlongs())
			djVassals.insert(std::make_pair(vassalID, nullptr));
	});
	registerKeyword("heir", [this](const std::string& unused, std::istream& theStream) {
		for (auto heirID: commonItems::llongList(theStream).getLlongs())
			heirs.emplace_back(std::make_pair(heirID, nullptr));
	});
	registerKeyword("laws", [this](const std::string& unused, std::istream& theStream) {
		const auto& theLaws = commonItems::stringList(theStream).getStrings();
		laws = std::set(theLaws.begin(), theLaws.end());
	});
	registerKeyword("holder", [this](const std::string& unused, std::istream& theStream) {
		holder = std::pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("coat_of_arms_id", [this](const std::string& unused, std::istream& theStream) {
		coa = std::pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("succession_election", [this](const std::string& unused, std::istream& theStream) {
		const auto newTitle = Title(theStream, 0);
		electors = newTitle.getElectors();
	});
	registerKeyword("electors", [this](const std::string& unused, std::istream& theStream) {
		for (auto electorID: commonItems::intList(theStream).getInts())
			electors.insert(std::pair(electorID, nullptr));
	});
	registerKeyword("landless", [this](const std::string& unused, std::istream& theStream) {
		landless = commonItems::singleString(theStream).getString() == "yes";
	});
	registerKeyword("color", [this](const std::string& unused, std::istream& theStream) {
		color = laFabricaDeColor.getColor(theStream);
	});
	registerKeyword("history", [this](const std::string& unused, std::istream& theStream) {
		previousHolders = Title(theStream, 0).getPreviousHolders();
	});
	registerRegex(R"(\d+.\d+.\d+)", [this](const std::string& unused, std::istream& theStream) {
		const auto questionableItem = commonItems::stringOfItem(theStream).getString();
		auto tempStream = std::stringstream(questionableItem);
		if (questionableItem.find('{') == std::string::npos)
		{
			try
			{
				previousHolders.emplace_back(std::pair(commonItems::singleLlong(tempStream).getLlong(), nullptr));
			}
			catch (std::exception&)
			{
				Log(LogLevel::Warning) << "Invalid previous holder ID: " << questionableItem;
			}
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void CK3::Title::cleanUpDisplayName()
{
	auto start = displayName.find('\x15');
	auto silence = false;
	auto silenceOverRide = false;
	auto braceDepth = 0;
	std::string newName;
	while (start != std::string::npos)
	{
		if ((!silence || silenceOverRide) && start != 0)
		{
			newName += displayName.substr(0, start);
			displayName = displayName.substr(start, displayName.length());
			silence = true;
			silenceOverRide = false;
			start = displayName.find('\x15');
		}
		else if (displayName.at(start + 1) == 'L')
		{
			if (displayName.length() >= start + 3)
				displayName = displayName.substr(start + 3, displayName.length());
			else
				break;
			silence = true;
			silenceOverRide = true;
			braceDepth++;
			start = displayName.find('\x15');
		}
		else if (displayName.at(start + 1) == '!')
		{
			if (displayName.length() >= start + 2)
				displayName = displayName.substr(start + 2, displayName.length());
			else
				break;
			braceDepth--;
			if (braceDepth <= 0)
			{
				silence = false;
				braceDepth = 0;
			}
			silenceOverRide = false;
			start = displayName.find('\x15');
		}
		else
		{
			silence = true;
			silenceOverRide = false;
			if (displayName.length() >= start + 1)
				displayName = displayName.substr(start + 1, displayName.length());
			else
				break;
			braceDepth++;
			start = displayName.find('\x15');
		}
	}
	displayName = newName;
}


int CK3::Title::flagDeJureHREProvinces()
{
	auto counter = 0;
	// For CK3 flagging dejure alone is not enough as we'll be losing a ton of land when duchies break away in full independence.
	// Flag our dejure AND defacto counties, or our dejure/defacto vassals' counties, wherever those may be.
	// Actual flagging will happen on county level as higher titles have no dejure provinces linked to them at this stage.
	// We're leaving baronies out of this, counties are sufficient.
	for (const auto& deJureVassal: djVassals)
	{
		counter += deJureVassal.second->flagDeJureHREProvinces();
	}
	for (const auto& deFactoVassal: dfVassals)
	{
		counter += deFactoVassal.second->flagDeJureHREProvinces();
	}
	if (!clay) // We really need the county clay to confirm we're a landful proper county.
		return counter;
	if (!clay->getCounty() || getLevel() != LEVEL::COUNTY)
		return counter;

	clay->getCounty()->second->setDeJureHRE();
	++counter;

	return counter;
}

void CK3::Title::brickTitle()
{
	grantIndependence();

	// Drop from own holder's domain.
	if (holder)
		holder->second->dropTitleFromDomain(ID);
	// Drop holder
	holder.reset();

	// release all vassals
	for (const auto& vassal: dfVassals)
		vassal.second->grantIndependence();
	dfVassals.clear(); // just in case?
}

void CK3::Title::dropTitleFromDFVassals(long long titleID)
{
	const auto& dfvItr = dfVassals.find(titleID);
	if (dfvItr != dfVassals.end())
		dfVassals.erase(dfvItr);
}

void CK3::Title::grantIndependence()
{
	// Drop this title from liege holder's vassals
	if (dfLiege)
		dfLiege->second->dropTitleFromDFVassals(ID);

	// Drop liege
	dfLiege.reset();
}

std::map<std::string, std::shared_ptr<CK3::Title>> CK3::Title::coalesceDFCounties() const
{
	// We're gathering vassal's counties + self (if c_), and passing them on, adding nothing to ourselves.
	std::map<std::string, std::shared_ptr<Title>> toReturn;
	for (const auto& vassal: dfVassals)
	{
		const auto& vassalCounties = vassal.second->coalesceDFCounties();
		toReturn.insert(vassalCounties.begin(), vassalCounties.end());
	}
	toReturn.insert(ownedDFCounties.begin(), ownedDFCounties.end());
	return toReturn;
}

std::map<std::string, std::shared_ptr<CK3::Title>> CK3::Title::coalesceDJCounties() const
{
	// We're gathering vassal dejure provinces + our own, and passing them on, adding nothing to ourselves.
	std::map<std::string, std::shared_ptr<Title>> toReturn;
	for (const auto& deJureVassal: djVassals)
	{
		const auto& vassalDJCounties = deJureVassal.second->coalesceDJCounties();
		toReturn.insert(vassalDJCounties.begin(), vassalDJCounties.end());
	}
	toReturn.insert(ownedDJCounties.begin(), ownedDJCounties.end());
	return toReturn;
}

void CK3::Title::congregateDFCounties()
{
	// We're gathering vassal counties and adding to our own.
	for (const auto& vassal: dfVassals)
	{
		const auto& vassalDFCounties = vassal.second->coalesceDFCounties();
		ownedDFCounties.insert(vassalDFCounties.begin(), vassalDFCounties.end());
	}
}

void CK3::Title::congregateDJCounties()
{
	// We're gathering de jure vassal de jure counties and adding to our own.
	for (const auto& deJureVassal: djVassals)
	{
		const auto& deJureVassalDJCounties = deJureVassal.second->coalesceDJCounties();
		ownedDJCounties.insert(deJureVassalDJCounties.begin(), deJureVassalDJCounties.end());
	}
}

CK3::LEVEL CK3::Title::getLevel() const
{
	// It's easy, until it's not.
	if (name.find("b_") == 0)
		return LEVEL::BARONY;
	if (name.find("c_") == 0)
		return LEVEL::COUNTY;
	if (name.find("d_") == 0)
		return LEVEL::DUCHY;
	if (name.find("k_") == 0)
		return LEVEL::KINGDOM;
	if (name.find("e_") == 0)
		return LEVEL::EMPIRE;

	// This is the questionable part. It should work for customs as they are treated identically as any other - with dejure parents and all.
	// exceptions are dynamic mercs, landless religious orders and similar - but those don't hold land, at least initially.

	// see if they hold any vassals and if so, assign a level one step higher.
	auto level = -1;
	const std::set<char> allowedPrefixes = {'b', 'c', 'd', 'k'};
	for (const auto& vassal: djVassals) // run through all as they can vary in levels.
	{
		if (!allowedPrefixes.count(vassal.second->getName().at(0)))
			continue; // Those dynamic vassals can't help us. Or heaven's forbid, it's an empire.
		if (LevelToInt[vassal.second->getLevel()] > level)
			level = LevelToInt[vassal.second->getLevel()] + 1;
	}
	if (level > -1)
		return IntToLevel[std::max(level, 4)];

	// Without vassals we must poke at our hierarchy, if any.
	if (djLiege)
		return IntToLevel[LevelToInt[djLiege->second->getLevel()] - 1];
	else
		return LEVEL::EMPIRE; // If this is wrong for landless mercs, it won't affect anything as they are landless anyway.
}

std::optional<commonItems::Color> CK3::Title::getColor() const
{
	if (color)
		return color;
	if (clay && clay->getColor())
		return clay->getColor();
	return std::nullopt;
}

bool CK3::Title::isLandless() const
{
	if (landless)
		return true;
	if (clay && clay->isLandless())
		return true;
	return false;
}

double CK3::Title::getBuildingWeight(const mappers::DevWeightsMapper& devWeightsMapper) const
{
	if (getLevel() != LEVEL::COUNTY) // This applies to nothing but counties.
		return 0;

	// buildingWeight is a mixture of all holdings, their potential buildings, and general county development.
	if (!clay)
		throw std::runtime_error("County " + name + " has no clay?");
	if (!clay->getCounty() || !clay->getCounty()->second)
		throw std::runtime_error("County " + name + " has no county in its clay?");
	const auto development = clay->getCounty()->second->getDevelopment();
	auto buildingCount = 0;
	auto holdingCount = 0;

	for (const auto& barony: djVassals)
	{
		if (!barony.second)
			throw std::runtime_error("Running unlinked vassals, are we? " + std::to_string(barony.first) + " has no link.");
		if (!barony.second->getClay())
			throw std::runtime_error("Supposed barony " + barony.second->getName() + " of " + name + " has no clay?");
		if (!barony.second->getClay()->getProvince() || !barony.second->getClay()->getProvince()->second)
			throw std::runtime_error("Barony " + barony.second->getName() + " of " + name + " has no clay province?");
		const auto& baronyProvince = barony.second->getClay()->getProvince();
		buildingCount += baronyProvince->second->countBuildings();
		if (!baronyProvince->second->getHoldingType().empty())
			++holdingCount;
	}

	const auto totalDev = devWeightsMapper.getDevFromHolding() * holdingCount + devWeightsMapper.getDevFromBuilding() * buildingCount +
								 devWeightsMapper.getDevFromDev() * development;
	return totalDev;
}

void CK3::Title::relinkDeFactoVassals()
{
	// We're reconstructing defacto hierarchy; we're redirecting our DFvassals from our DFtitle into our DJtitle - if holder of this title holds hold both.
	// This is important - otherwise secondary titles wouldn't have their own vassals under them but under primary title, breaking PU splitoffs,
	// and resulting with those secondary duchies/kingdoms being discarded as landless.

	if (name.starts_with("c_") || name.starts_with("b_"))
		return; // don't bother with counties and below.

	for (const auto& dfVassal: dfVassals)
	{
		if (!dfVassal.second)
			continue;

		if (dfVassal.second->getDFLiege() && dfVassal.second->getDJLiege() && dfVassal.second->getDFLiege()->first != dfVassal.second->getDJLiege()->first)
		{
			// Do we own both titles?
			if (holder && holder->second && holder->second->getDomain())
			{
				const auto holderTitles = holder->second->getDomain()->getDomain();

				// Make a small cache of the holder's owned title names.
				std::set<std::string> titleNameCache;
				for (const auto& domainTitle: holderTitles)
					titleNameCache.insert(domainTitle.second->getName());

				if (titleNameCache.contains(dfVassal.second->getDFLiege()->second->getName()) &&
					 titleNameCache.contains(dfVassal.second->getDJLiege()->second->getName()))
				{
					// We do own both. Tell the vassal to relink its defacto owner to dejure title.
					dfVassal.second->loadDFLiege(*dfVassal.second->getDJLiege());
					// Tell dejure title it's now owner of a brand new vassal.
					dfVassal.second->getDJLiege()->second->addDFVassals(std::map{dfVassal});
					// And finally...
					dropTitleFromDFVassals(dfVassal.first);
				}
			}
			else
			{
				if (!holder)
					Log(LogLevel::Warning) << name << " has no holder but has vassals?";
				else if (!holder->second)
					Log(LogLevel::Warning) << name << "'s holder is not linked up!";
				else if (!holder->second->getDomain())
					Log(LogLevel::Warning) << name << "' holder has no domain but has vassals?";
			}
		}
	}
}

File: CK3ToEU4/Source/CK3/Titles/Title.cs
using System;
using System.Collections.Generic;
using CK3ToEU4.CK3.CoatsOfArms;
using CK3ToEU4.Configuration;
using commonItems;
using commonItems.Colors;

namespace CK3ToEU4.CK3.Titles;

enum Level
{
    Barony,
    County,
    Duchy,
    Kingdom,
    Empire
} 

public class Title
{
	public long Id { get; } = 0; // 11038
	public bool isTheocraticLease() { return theocraticLease; }
	public bool isCountyCapitalBarony() { return cCapitalBarony; }
	public bool isDuchyCapitalBarony() { return dCapitalBarony; }
	public bool isHREEmperor() { return HREEmperor; }
	public bool isInHRE() { return inHRE; }
	public bool isThePope() { return thePope; }
	public bool isElectorate() { return electorate; }
	public bool isHolderCapital() { return holderCapital; }
	public bool isHRECapital() { return HRECapital; }
	public bool isCustomTitle() { return customTitle; }
	public bool isRenamed() { return renamed; }
	public bool isManualNameClaimed() { return nameClaimed; }
	public string getName() { return name; }
	public string getDisplayName() { return displayName; }
	public string getAdjective() { return adjective; }
	public Date getCreationDate() { return creationDate; }
	public string getHistoryGovernment() { return historyGovernment; }
	public KeyValuePair<long, Title> getCapital() { return capital; }
	public KeyValuePair<long, Title>? getDFLiege() { return dfLiege; }
	public KeyValuePair<long, Title> getDJLiege() { return djLiege; }
	public Dictionary<long, Title> getDFVassals() { return dfVassals; }
	public Dictionary<long, Title> getDJVassals() { return djVassals; }
	public List<KeyValuePair<long, Character>> getHeirs() { return heirs; }
	public Dictionary<long, Character> getClaimants() { return claimants; }
	public Dictionary<long, Character?> getElectors() { return electors; }
	public HashSet<string> getLaws() { return laws; }
	public KeyValuePair<long, Character>? getHolder() { return holder; }
	public KeyValuePair<long, CoatOfArms>? getCoA() { return coa; }
	public LandedTitles getClay() { return clay; }
	public Dictionary<string, Title> getOwnedDFCounties() { return ownedDFCounties; } // TODO: change return type
	public Dictionary<string, Title> getOwnedDJCounties() { return ownedDJCounties; }
	public KeyValuePair<string, Title> getHoldingTitle() { return holdingTitle; }
	public KeyValuePair<string, Title>? getGeneratedLiege() { return generatedLiege; }
	public Dictionary<string, Title> getGeneratedVassals() { return generatedVassals; }
	public KeyValuePair<string, Country>? getEU4Tag() { return tagCountry; }
	public List<KeyValuePair<long, Character>> getPreviousHolders() { return previousHolders; }

	public Color? getColor();
	public bool isLandless();
	public bool areDFLiegeAndDJLiegeSet();
	public bool areDFLiegeAndDJLiegeSame();
	public bool isHolderSet() { return holder != std::nullopt; }
	public bool isHolderLinked() { return holder  holder->second; }
	public bool doesHolderHaveCharacterDomain();
	public HashSet<string> getTitleNamesFromHolderDomain();
	public Level getLevel();

	public Title(BufferedReader reader, long theID)
	{
		Id = theID;

		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
		
		// Obligatory sanity check.
		if (!holder)
		{
			// This title does NOT in fact exist.
			dfLiege = null;
		}
	}

	#region linkage
	public void loadCoat(KeyValuePair<long, CoatOfArms> coat) { coa = coat; }
	public void loadDFLiege(KeyValuePair<long, Title> DFLiege) { dfLiege = DFLiege; }
	public void loadDJLiege(KeyValuePair<long, Title> DJLiege) { djLiege = DJLiege; }
	public void loadDFVassals(Dictionary<long, Title> DFVassals) { dfVassals = DFVassals; }
	public void addDFVassals(Dictionary<long, Title> DFVassals) { dfVassals.insert(DFVassals.begin(), DFVassals.end()); }
	public void loadDJVassals(Dictionary<long, Title> DJVassals) { djVassals = DJVassals; }
	public void loadHolder(KeyValuePair<long, Character> theHolder) { holder = theHolder; }
	public void loadHeirs(List<KeyValuePair<long, Character>> theHeirs) { heirs = theHeirs; }
	public void loadPreviousHolders(List<KeyValuePair<long, Character>> theHolders) { previousHolders = theHolders; }
	public void loadClaimants(Dictionary<long, Character> theClaimants) { claimants = theClaimants; }
	public void loadElectors(Dictionary<long, Character> theElectors) { electors = theElectors; }
	public void loadClay(LandedTitles theClay) { clay = theClay; }
	public void loadOwnedDFCounties( Dictionary<string, Title> theOwnedCounties) { ownedDFCounties = theOwnedCounties; }
	public void loadOwnedDJCounties( Dictionary<string, Title> theOwnedCounties) { ownedDJCounties = theOwnedCounties; }
	#endregion

	#region processing
	public void setDynamicLevel(Level theLevel) { dynamicLevel = theLevel; }
	public void cleanUpDisplayName();
	public int flagDeJureHREProvinces();
	public void brickTitle();
	public void grantIndependence();
	public void resetDFLiege() { dfLiege.reset(); }
	public void resetDJLiege() { djLiege.reset(); }
	public void setHREEmperor() { HREEmperor = true; }
	public void setInHRE() { inHRE = true; }
	public void dropTitleFromDFVassals(long long titleID);
	public void setThePope() { thePope = true; }
	public void setCustomTitle() { customTitle = true; }
	public void setManualNameClaim() { nameClaimed = true; }
	public void pickDisplayName(const Dictionary<string, Title?>& possibleTitles); // Grants one county's name to another during N:1/N:M mappings
	public Title findDuchyCapital();																 // Only for c_, for now
	public void congregateDFCounties();
	public void congregateDJCounties();
	public void loadGeneratedLiege(const KeyValuePair<string, Title?>& liege) { generatedLiege = liege; }
	public void addGeneratedVassal(const KeyValuePair<string, Title?>& theVassal) { generatedVassals.insert(theVassal); }
	public void loadHoldingTitle(const KeyValuePair<string, Title?>& theTitle) { holdingTitle = theTitle; }
	public void setElectorate() { electorate = true; }
	public void relinkDeFactoVassals();
	#endregion

	#region conversion
	public void loadEU4Tag(const KeyValuePair<string, std::shared_ptr<EU4::Country>>& theCountry) { tagCountry = theCountry; }
	public double getBuildingWeight(const mappers::DevWeightsMapper& devWeightsMapper) const;
	public void setHolderCapital() { holderCapital = true; }
	public void setHRECapital() { HRECapital = true; }
	public void clearGeneratedVassals() { generatedVassals.clear(); }

	Dictionary<string, Title> coalesceDFCounties() const;
	Dictionary<string, Title> coalesceDJCounties() const;
	#endregion

	private void RegisterKeys(Parser parser, ColorFactory colorFactory)
	{
			parser.RegisterKeyword("key", reader => {
		name = reader.GetString();
	});
	parser.RegisterKeyword("name", reader => {
		displayName = reader.GetString();
		if (displayName.find("\x15") != string::npos)
		{
			cleanUpDisplayName();
		}
	});
	parser.RegisterKeyword("adj", reader => {
		adjective = reader.GetString();
	});
	parser.RegisterKeyword("date", reader => {
		creationDate = reader.GetString();
	});
	parser.RegisterKeyword("claim", reader => {
		foreach (var claimantID in reader.GetLongs())
			claimants.Add(claimantID, null);
	});
	parser.RegisterKeyword("history_government", reader => {
		historyGovernment = reader.GetString();
	});
	parser.RegisterKeyword("theocratic_lease", reader => {
		theocraticLease = reader.GetString() == "yes";
	});
	parser.RegisterKeyword("capital_barony", reader => {
		cCapitalBarony = reader.GetString() == "yes";
	});
	parser.RegisterKeyword("duchy_capital_barony", reader => {
		dCapitalBarony = reader.GetString() == "yes";
	});
	parser.RegisterKeyword("capital", reader => {
		capital = new(reader.GetLong(), null);
	});
	parser.RegisterKeyword("de_facto_liege", reader => {
		dfLiege = KeyValuePair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	parser.RegisterKeyword("de_jure_liege", reader => {
		djLiege = KeyValuePair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	parser.RegisterKeyword("de_jure_vassals", reader => {
		for (auto vassalID: commonItems::llongList(theStream).getLlongs())
			djVassals.insert(std::make_pair(vassalID, nullptr));
	});
	parser.RegisterKeyword("heir", reader => {
		for (auto heirID: commonItems::llongList(theStream).getLlongs())
			heirs.emplace_back(std::make_pair(heirID, nullptr));
	});
	parser.RegisterKeyword("laws", reader => {
		const auto& theLaws = commonItems::stringList(theStream).getStrings();
		laws = std::set(theLaws.begin(), theLaws.end());
	});
	parser.RegisterKeyword("holder", reader => {
		holder = KeyValuePair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	parser.RegisterKeyword("renamed", [this](std::istream& theStream) {
		renamed = commonItems::getString(theStream) == "yes";
	});
	parser.RegisterKeyword("coat_of_arms_id", reader => {
		coa = new KeyValuePair<long, CoatOfArms?>(reader.GetLong(), null);
	});
	parser.RegisterKeyword("succession_election", reader => {
		const auto newTitle = Title(theStream, 0);
		electors = newTitle.getElectors();
	});
	parser.RegisterKeyword("electors", reader => {
		foreach (var electorID in reader.GetInts())
			electors.Add(new KeyValuePair<long, Character>(electorID, null));
	});
	parser.RegisterKeyword("landless", reader => {
		landless = reader.GetString() == "yes";
	});
	parser.RegisterKeyword("color", reader => {
		color = colorFactory.GetColor(reader);
	});
	parser.RegisterKeyword("history", reader => {
		previousHolders = new Title(reader, 0).getPreviousHolders();
	});
	parser.RegisterRegex(CommonRegexes.Date, reader =>
	{
		var questionableItem = reader.GetStringOfItem().ToString();
		var tempReader = new BufferedReader(questionableItem);
		if (!questionableItem.Contains('{'))
		{
			try
			{
				previousHolders.Add(new(tempReader.GetLong(), null));
			}
			catch (Exception e)
			{
				Logger.Warn($"Invalid previous holder ID: {questionableItem}");
			}
		}
	});
	parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}


    																			
    private KeyValuePair<long, Title?> capital;							// capital title is a COUNTY, even for county itself and baronies beneath it!
    private string name;																			// c_ashmaka
    private string displayName;																// Ashmaka
    private string adjective;																	// Ashmakan
    private string article;																		// "the ". Not always present.
    private string historyGovernment;														// Unclear why this is history. Maybe further governments override it.
    private Date creationDate;																		// Unclear. Ranges to 9999.1.1, probably is PDX alternative for "bool isCreated";
    private KeyValuePair<long, CoatOfArms?>? coa;	// This is dejure flag but not defacto.
    private KeyValuePair<long, Title>? dfLiege;		// defacto liege title (d_kalyani)
    private KeyValuePair<long, Title>? djLiege;		// dejure liege title (d_rattapadi)
    private KeyValuePair<long, Character>? holder; // Holding character
    private Dictionary<long, Title> dfVassals;						// defacto vassals, not in save, manually linked post-loading
    private Dictionary<long, Title> djVassals;						// dejure vassals (for all except baronies and titulars)
    private List<KeyValuePair<long, Character>> heirs;		// Order of heirs is unclear so we're keeping them ordered and using first if able.
    private Dictionary<long, Character?> claimants;					// People holding a claim to this title. Incredibly useful.
    private Dictionary<long, Character> electors;					// People involved in elections regardless of election type law.
    private bool theocraticLease = false;															// Does this apply to non-baronies? Maybe? Who owns it then, dejure liege?
    private HashSet<string> laws;
    private bool cCapitalBarony = false;
    private bool dCapitalBarony = false;
	private LandedTitles clay; // Middleware towards geographical data, essential for b_c_, potentially obsolete for others.
	private bool HREEmperor = false;
	private bool inHRE = false;
	private bool thePope = false;
	private bool customTitle = false;													// True if title was fromed via "Found a New Kingdom/Empire" decision. Vestigal
	private bool renamed = false;														// True if title was manually named
	private bool nameClaimed = false;													// Flag for 1:M (or N:M) province mappings to decide who gets the manual name
	private Dictionary<string, Title> ownedDFCounties; // used to map higher-lvl titles directly to clay. Includes self! Every c_+ title has this.
	private Dictionary<string, Title> ownedDJCounties; // ditto
	private KeyValuePair<string, Title>? generatedLiege; // Liege we set manually while splitting vassals.
	private Dictionary<string, Title> generatedVassals;					// Vassals we split off deliberately.
	private KeyValuePair<string, Title> holdingTitle;						// topmost owner title (e_francia or similar), only c_s have this.
	private bool electorate = false;
	private KeyValuePair<string, EU4.Country>? tagCountry;
	private bool landless = false;
	private Color? color;
	private List<KeyValuePair<long, Character?>> previousHolders;
	private bool holderCapital = false;
	private bool HRECapital = false;
	private Level? dynamicLevel; // Maybe assigned through dynamic ranks, otherwise getLevel will try to guesstimate.

}